### ISSUE TITLE
switch build scripts to use go modules, stop requiring $GOPATH

### ISSUE THREAD

**Issue by liggitt** [kubernetes/kubernetes#82531]
Tasks to move to go1.13:

- [x] resolve https://github.com/kubernetes/kubernetes/issues/82506 / https://github.com/heketi/heketi/issues/1648 (fixed upstream https://github.com/heketi/heketi/pull/1649)
  - fixed in k8s in #82805
- [x] update gofmt-formatted source to be compatible with 1.12 and 1.13
  - https://github.com/kubernetes/kubernetes/pull/82684
- [x] go1.13.0 scalability regression (golang/go#32828)
- [x] http connection caching regression (https://github.com/golang/go/issues/34978)
- [x] update golang.org/x/... dependencies to go1.13 branches
- [x] update build to use GOPATH short-term
  - WIP in https://github.com/kubernetes/kubernetes/pull/82809

Things that need to happen before $GOPATH support is dropped (potentially go1.18?):
- [x] fix code-generator/gengo vendor assumptions - https://github.com/kubernetes/kubernetes/issues/86753
- [x] fix excessive `go list` use in build scripts causing extremely long build times
- [x] rework go2make to fix performance issue (takes 2-3 seconds with $GOPATH, several minutes with modules) - https://github.com/golang/go/issues/31087#issuecomment-761692033
- [x] remove $GOPATH requirements and fake `_output` $GOPATH setup from build code (build with modules+vendor)

/cc @fejta @BenTheElder @cblecker 
/sig release

**Comment by aanm** [comment]
>fix excessive go list use in build scripts causing extremely long build times in go1.13 (go list is fast in go1.12, slow in go1.13)

@liggitt The only way I found out to keep using go list without long build times was to run with `GO111MODULE=off` for go lint commands.

**Comment by liggitt** [comment]
I think @thockin was looking into running the command once and caching/reusing the results for the remainder of  the `make` invocation

**Comment by bcmills** [comment]
> `go list` is fast in go1.12, slow in go1.13

That may be a side-effect of the fix for golang/go#29773, since it seems that you do have a lot of dependencies through older versions from the main module (due to heketi/heketi#1648).

If the `go list` performance regression you're seeing is not attributable to that, then please do file an issue — I'm not aware of anything else that should have slowed it down at all.

**Comment by liggitt** [comment]
Hmm, maybe it wasn't that `go list` was slower, it was that it started having side effects, and the side-effect-free alternatives were slower. xref https://github.com/golang/go/issues/31087

**Comment by liggitt** [comment]
heketi/heketi issue resolved in https://github.com/heketi/heketi/pull/1649, vendored copy bumped in https://github.com/kubernetes/kubernetes/pull/82805

**Comment by thockin** [comment]
The things that call go list have to be totally rewritten.

https://github.com/golang/go/issues/31087

**Comment by thockin** [comment]
compiling with 1.13 gives me:

```
$ make kubectl
go/build: importGo k8s.io/kubernetes: exit status 1
can't load package: package k8s.io/kubernetes: cannot find module providing package k8s.io/kubernetes


+++ [0917 16:03:22] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
go: finding k8s.io/kubernetes/vendor latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen latest
go: finding k8s.io/kubernetes/vendor/k8s.io latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd latest
can't load package: package k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen: no matching versions for query "latest"
!!! [0917 16:03:24] Call tree:
!!! [0917 16:03:24]  1: /usr/local/google/home/thockin/src/k/kubernetes/hack/lib/golang.sh:714 kube::golang::build_some_binaries(...)
!!! [0917 16:03:24]  2: /usr/local/google/home/thockin/src/k/kubernetes/hack/lib/golang.sh:853 kube::golang::build_binaries_for_platform(...)
!!! [0917 16:03:24]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0917 16:03:24] Call tree:
!!! [0917 16:03:24]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0917 16:03:24] Call tree:
!!! [0917 16:03:24]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [Makefile.generated_files:200: _output/bin/deepcopy-gen] Error 1
make: *** [Makefile:559: generated_files] Error 2
```

At the bottom it calls `go install k8s.io/kubernetes/./vendor/k8s.io/code-generator/cmd/deepcopy-gen`.  I forget exactly why that prefix was needed but it was.  If I remove it, it passes.

**Comment by thockin** [comment]
...but the build fails later:

```
$ make kubectl
go/build: importGo k8s.io/kubernetes: exit status 1
can't load package: package k8s.io/kubernetes: cannot find module providing package k8s.io/kubernetes


+++ [0917 16:05:28] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
go: downloading k8s.io/klog v0.4.0
go: downloading k8s.io/gengo v0.0.0-20190822140433-26a664648505
go: extracting k8s.io/klog v0.4.0
go: extracting k8s.io/gengo v0.0.0-20190822140433-26a664648505
go: downloading golang.org/x/tools v0.0.0-20190621195816-6e04913cbbac
go: extracting golang.org/x/tools v0.0.0-20190621195816-6e04913cbbac
go: finding k8s.io/gengo v0.0.0-20190822140433-26a664648505
go: finding k8s.io/klog v0.4.0
go: finding golang.org/x/tools v0.0.0-20190621195816-6e04913cbbac
F0917 16:06:34.716074  242494 main.go:82] Error: Failed making a parser: unable to add directory "k8s.io/kubernetes/vendor/k8s.io/api/admission/v1": unable to import "k8s.io/kubernetes/vendor/k8s.io/api/admission/v1": go/build: importGo k8s.io/kubernetes/vendor/k8s.io/api/admission/v1: exit status 1
go: finding k8s.io/kubernetes/vendor/k8s.io latest
go: finding k8s.io/kubernetes/vendor latest
go: finding k8s.io/kubernetes/vendor/k8s.io/api latest
go: finding k8s.io/kubernetes/vendor/k8s.io/api/admission latest
can't load package: package k8s.io/kubernetes/vendor/k8s.io/api/admission/v1: no matching versions for query "latest"

make[1]: *** [Makefile.generated_files:152: gen_deepcopy] Error 1
make: *** [Makefile:559: generated_files] Error 2
```

**Comment by fejta** [comment]
We use modules, golang1.13 (and bazel) in test-infra. The way we found to get the generators to work is to temporarily disable modules:

https://github.com/kubernetes/test-infra/blob/master/hack/update-codegen.sh#L119

Here's how kind does the same thing without bazel's help:

https://github.com/kubernetes-sigs/kind/blob/master/hack/update/generated.sh

Does `GO111MODULE=off` change anything?

Are we sure we need to generate code for people on the fly? We could instead ask people to call go generate explicitly when they want to regenerate things.

**Comment by thockin** [comment]
if we depend on people calling `go generate` they WILL run tests without having regenerated code.  This is why we did on-the-fly in the first place

**Comment by fejta** [comment]
We can have presubmits catch this though, right?

**Comment by thockin** [comment]
What a miserable experience, but I guess, maybe?

**Comment by fejta** [comment]
This is how test-infra handles things. We don't generate for people, but presubmits will fail if they forget. 

It seems like go is pretty committed to making people use this pattern? It seems hard to reconcile doing anything else with the expectation that you can `go get` code.

For example historically we generated protos on the fly with bazel, but it caused enough annoyance that we abandoned it in favor of an update-protos.sh (and verify-protos.sh) script and checking in the .pb.go files... Paradoxically people seem a lot happier since this change, despite the need to update manually.

**Comment by liggitt** [comment]
> At the bottom it calls go install k8s.io/kubernetes/./vendor/k8s.io/code-generator/cmd/deepcopy-gen. I forget exactly why that prefix was needed but it was. If I remove it, it passes.

That's where I ended up in https://github.com/liggitt/kubernetes/commits/go-1.13 as well

**Comment by liggitt** [comment]
The immediately motivation to get into go1.13 is for support reasons. Supported go versions last ~1 year, we're six months into go1.12, and we're about to ship k8s 1.16 and support it for ~9 months. My ideal scenario is to get a small set of changes in build/generation scripts we can pick to release-1.16 to get it into a go version supported for its lifetime.

Long term, I'm hopeful the move to go1.13 and to using modules in build scripts will let us clean out a lot of GOPATH cruft we've accumulated, though I'd expect that cleanup to be more invasive and hopefully limited to k8s 1.17 only.

**Comment by liggitt** [comment]
looks like simply setting GO111MODULE=off in init.sh works around the build issues... that could be a reasonable first step to validate go1.13, and consider picking back to 1.16 (not urgently, just sometime in the next few months)

**Comment by cblecker** [comment]
1.13.1 is now out

**Comment by bcmills** [comment]
@cblecker, note that 1.13.1 was a security release. The backports for performance regressions are slated for 1.13.2.

**Comment by tpepper** [comment]
1.13.2 is coming https://groups.google.com/forum/#!topic/golang-announce/7zbgLB1kCKU in two days

**Comment by BenTheElder** [comment]
This is also a security release, have all the relevant non-security patches
landed?

**Comment by liggitt** [comment]
> This is also a security release, have all the relevant non-security patches landed?

If it follows the pattern of 1.13.1, only the security patch will be included, and we will have to wait for 1.13.3

**Comment by bcmills** [comment]
Correct, security releases include only security-impacting changes.

**Comment by mikedanese** [comment]
We are hitting another issue with connection caching in http.Transport. Will fill in details in a bit.

**Comment by liggitt** [comment]
Specific to (or worse in) 1.13?

**Comment by mikedanese** [comment]
Yes, I currently believe so.

**Comment by mikedanese** [comment]
Looks like this is broken:

https://github.com/golang/go/commit/43b9fcf6fe9feb0ec67a7ddb01e5b542542b47c8

**Comment by mikedanese** [comment]
Looks like this is broken:

https://github.com/golang/go/commit/43b9fcf6fe9feb0ec67a7ddb01e5b542542b47c8

I filed https://github.com/golang/go/issues/34978

**Comment by zhangxiaoyu-zidif** [comment]
I checkout release-1.16 and execute `make`, and terrible experience with 
```shell
go: finding github.com/spf13/pflag v1.0.3
go/build: importGo k8s.io/kubernetes: exit status 1
can't load package: package k8s.io/kubernetes: cannot find module providing package k8s.io/kubernetes


+++ [1128 17:29:18] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator latest
go: finding k8s.io/kubernetes/vendor latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen latest
go: finding k8s.io/kubernetes/vendor/k8s.io latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd latest
can't load package: package k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen: no matching versions for query "latest"
!!! [1128 17:30:00] Call tree:
!!! [1128 17:30:00]  1: /Users/zidif/Downloads/gopath/src/k8s.io/kubernetes/hack/lib/golang.sh:714 kube::golang::build_some_binaries(...)
!!! [1128 17:30:00]  2: /Users/zidif/Downloads/gopath/src/k8s.io/kubernetes/hack/lib/golang.sh:853 kube::golang::build_binaries_for_platform(...)
!!! [1128 17:30:00]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [_output/bin/deepcopy-gen] Error 1
make: *** [generated_files] Error 2
```

**Comment by zhangxiaoyu-zidif** [comment]
> I checkout release-1.16 and execute `make`, and terrible experience with
> 
> ```shell
> go: finding github.com/spf13/pflag v1.0.3
> go/build: importGo k8s.io/kubernetes: exit status 1
> can't load package: package k8s.io/kubernetes: cannot find module providing package k8s.io/kubernetes
> 
> 
> +++ [1128 17:29:18] Building go targets for darwin/amd64:
>     ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
> go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator latest
> go: finding k8s.io/kubernetes/vendor latest
> go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen latest
> go: finding k8s.io/kubernetes/vendor/k8s.io latest
> go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd latest
> can't load package: package k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen: no matching versions for query "latest"
> !!! [1128 17:30:00] Call tree:
> !!! [1128 17:30:00]  1: /Users/zidif/Downloads/gopath/src/k8s.io/kubernetes/hack/lib/golang.sh:714 kube::golang::build_some_binaries(...)
> !!! [1128 17:30:00]  2: /Users/zidif/Downloads/gopath/src/k8s.io/kubernetes/hack/lib/golang.sh:853 kube::golang::build_binaries_for_platform(...)
> !!! [1128 17:30:00]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
> make[1]: *** [_output/bin/deepcopy-gen] Error 1
> make: *** [generated_files] Error 2
> ```
Resolved: 
1.12.9 <= go version < 1.13

**Comment by tedyu** [comment]
For go version go1.13.6 linux/amd64
```
go: finding k8s.io/kubernetes/vendor/k8s.io latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen latest
go: finding k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd latest
can't load package: package k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/deepcopy-gen: no matching versions for query "latest"
!!! [0315 14:18:14] Call tree:
!!! [0315 14:18:14]  1: /home/zyu/kubernetes/hack/lib/golang.sh:630 kube::golang::build_some_binaries(...)
!!! [0315 14:18:14]  2: /home/zyu/kubernetes/hack/lib/golang.sh:769 kube::golang::build_binaries_for_platform(...)
!!! [0315 14:18:14]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0315 14:18:14] Call tree:
!!! [0315 14:18:14]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0315 14:18:14] Call tree:
!!! [0315 14:18:14]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
```
Here is the OS:
```
Linux desktop 5.5.7-200.fc31.x86_64 #1 SMP Fri Feb 28 17:18:37 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by BenTheElder** [comment]
@liggitt pointed out as of 1 hour ago this is now something we _have_ to do, go will drop support for GOPATH mode at some point https://github.com/golang/go/issues/37755#issuecomment-780781769

**Comment by spiffxp** [comment]
/priority important-longterm
/area release-eng
FYI @kubernetes/release-engineering

**Comment by liggitt** [comment]
Looks like removal of GOPATH mode is targeting go1.17 per https://blog.golang.org/go116-module-changes. This just got more important.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/94227
https://github.com/kubernetes/kubernetes/issues/86753

Did some minor cleanup in https://github.com/kubernetes/kubernetes/pull/99923

**Comment by BenTheElder** [comment]
/remove-lifecycle stale

tentatively it looks like we may not have to do this by 1.17 but we probably still need to long term.

https://tip.golang.org/doc/go1.17

> vendor contents
>
> If the main module specifies go 1.17 or higher, go mod vendor now annotates vendor/modules.txt with the go version indicated by each vendored module in its own go.mod file. The annotated version is used when building the module's packages from vendored source code.
>
> If the main module specifies go 1.17 or higher, go mod vendor now omits go.mod and go.sum files for vendored dependencies, which can otherwise interfere with the ability of the go command to identify the correct module root when invoked within the vendor tree.

**Comment by voigt** [comment]
Hi there,
Bug-Triage here: given the current state of v1.22 release, does it make sense to move this issue to `milestone v1.23` (or even clear the milestone)?

**Comment by liggitt** [comment]
/sig api-machinery

/assign @thockin
for go2make

/assign @sttts
for code generation

**Comment by thockin** [comment]
If they force GO111MODULE to be ignored, our builds will slow down by a
TON.

https://groups.google.com/d/msgid/golang-nuts/CAO_Rewa6rMW79iBHj2Jz6HfJ-tCFLFhNAhYiwDh%3DNy6M35Y91Q%40mail.gmail.com

**Comment by liggitt** [comment]
> If they force GO111MODULE to be ignored, our builds will slow down by a TON.

I'd take slow-but-correct as a starting point, but right now, the gengo/code-generation bits fail if you try to build with GO111MODULE=on (their internal logic about package paths, vendoring, etc, do not hold)

**Comment by BenTheElder** [comment]
We now have confirmation that go 1.17 at least will not drop non module mode.

**Comment by thockin** [comment]
gengo does make some dubious assumptions.

**Comment by thockin** [comment]
From the linked thread: "Despite what the blog post says, Go 1.17 won't
drop support for GO111MODULE or GOPATH mode. It will be more or less the
same as GOPATH mode."

I think we can count on them not dumping it until our class of problems is
dealt with.

**Comment by BenTheElder** [comment]
xref: https://github.com/kubernetes/kubernetes/pull/105891

**Comment by jyotimahapatra** [comment]
Hi 👋 . Checking in from bug triage for 1.24. Should we be tracking progress on this issue for 1.24 release?

**Comment by jyotimahapatra** [comment]
Hi @liggitt 👋 . I'm checking in from bug triage team. Is this issue being targeted for 1.24 milestone?

**Comment by liggitt** [comment]
no, will move to 1.25. go1.18 added support for go workspaces, which might be helpful here (@thockin is carving up the make logic gradually, but that will not complete in 1.24)

/milestone v1.25

**Comment by thockin** [comment]
Hopefully I can get more time soon so we can get this in for 25

**Comment by encodeflush** [comment]
Hi @liggitt 
My name Hossein Salahi from Bug-Triage team. I am just checking to ensure if everything is on track for K8s 1.25.
Thanks!

**Comment by liggitt** [comment]
I doubt it will be completed in 1.25, but some work is happening in the 1.25 timeframe (https://groups.google.com/a/kubernetes.io/g/dev/c/zcGFMru0gPU/m/h1OlbdBtAwAJ), so the milestone is still appropriate

**Comment by helayoty** [comment]
@liggitt 👋 the release 1.25 bug triage Lead. I'm reaching out to see the status of this Issue and if we are still targeting the current release.

**Comment by liggitt** [comment]
resolved by https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4402-go-workspaces
